refactor(core): Move backend config to a separate package (no-changelog)

netroy opened the pull request:
This PR moves the config related code into a separate `@n8n/config` package so that we can use the config on all backend packages (like `core` and `nodes-base`) without creating a circular dependency with `cli`.
This PR also changes the config object to be
1. an injectable class, that can be used with regular DI code
2. type-safe without the need for the complicated inference code in `packages/cli/node_modules/@types/convict/index.d.ts`
3. annotated with JSDocs on every config property, which adds helpful tooltips in our editors
![image](https://github.com/n8n-io/n8n/assets/196144/67e76ee3-3586-48ce-92f5-98acf997daa4)

Every sub-section in the config is also now separately injectable, which we can use to create config objects specific to certain sections of the codebase, instead of using this one global config everywhere.

# Ticket
ENG-126 & ENG-126

## Review / Merge checklist
- [x] PR title and summary are descriptive
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [x] Tests included

netroy committed:
move backend config to a separate package

ivov commented on packages/@n8n/config/src/configs/database.ts:
```diff
@@ -0,0 +1,159 @@
+@Config
+class LoggingConfig {
+	/** Typeorm logging enabled flag. @default false */
+	@FromEnv('DB_LOGGING_ENABLED')
```
Subjective, I'd prefer shorter syntax:

```suggestion
	@Env('DB_LOGGING_ENABLED')
```

ivov commented on packages/@n8n/config/src/configs/database.ts:
```diff
@@ -0,0 +1,159 @@
+
+@Config
+class LoggingConfig {
+	/** Typeorm logging enabled flag. @default false */
```
I tend to think `@default` looks redundant and liable to drift. With IDE navigation we can find this out quickly now.

ivov commented on packages/@n8n/config/src/configs/database.ts:
```diff
@@ -0,0 +1,159 @@
+class LoggingConfig {
+	/** Typeorm logging enabled flag. @default false */
+	@FromEnv('DB_LOGGING_ENABLED')
+	readonly enabled: boolean = false;
```
Ideally there should be a way to have the effect of `readonly` but without overly narrowing the type and without having to specify it... Not sure if possible but I'll look if there's a way.

ivov replied:
Not possible :/

ivov commented on packages/@n8n/config/src/configs/database.ts:
```diff
@@ -0,0 +1,159 @@
+	readonly enabled: boolean = false;
+
+	/**
+	 * Logging level options, default is "error".
```
Else we specify it 3 times:

```suggestion
	 * Logging level options.
```

ivov commented on packages/@n8n/config/.eslintrc.js:
```diff
@@ -0,0 +1,15 @@
+
+	rules: {
+		'@typescript-eslint/consistent-type-imports': 'error',
+		'n8n-local-rules/no-plain-errors': 'off',
```
Why turn this off?

ivov commented on packages/@n8n/config/src/decorators.ts:
Love how little code is needed and how clear it is 👍🏻

ivov commented on packages/@n8n/config/src/index.ts:
```diff
@@ -0,0 +1,8 @@
+import { DatabaseConfig } from './configs/database';
+
+@Config
+export class GlobalConfig {
+	@Nested
+	database: DatabaseConfig;
+}
```
We might consider utility methods, e.g. `restoreDefaults()` for tests.

netroy replied:
to restore defaults for tests, we can just create a new instance of the class.

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
address PR comments

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
use destructuring to retain JSDocs wherever possible

netroy committed:
add support for reading individual config value from files defined using `_FILE` suffixed env variables

netroy committed:
move credentials config over

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
move userManagement.emails config over

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

ivov reviewed (commented):
Submitting for now but not finished

TODOs:
- [x] Manually test envs → pass
- [x] Manually test `N8N_CONFIG_FILES` → ~fail~ pass
- [x] Manually test `_FILE` → pass
- [ ] Look closer at new decorators
- [ ] Look closer at mailer changes

netroy commented on packages/@n8n/config/src/configs/email.ts:
```diff
@@ -0,0 +1,78 @@
+
+@Config
+export class EmailConfig {
+	/** How to send emails */
```
almost all the comments here are [literally copy-pasted from what we have in `schema.ts`](https://github.com/n8n-io/n8n/blob/master/packages/cli/src/config/schema.ts#L819). We should definitely improve all these docs, but I don't think we need to block this PR on these docs.
It might make sense to get the docs team involved once the new package has all the config variables in it.

netroy committed:
Apply suggestions from code review

Co-authored-by: Iván Ovejero <ivov.src@gmail.com>

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy commented:
Addressed most of the comments. But overall, I don't think this PR should be fixing the existing docs for these config variables. We need a more involved overhaul of these docs, collaborating with the docs team.

Regarding the mailer changes, those could be a separate PR. Please let me know if you prefer it that way 🙏🏽

ivov commented:
Tested `export N8N_CONFIG_FILES=$N8N_CONFIG_FILES_DIR/local-postgres.json` where the file is as below, and I'm seeing the following error. This does not fail on master.

```json
{
  "database": {
    "type": "postgresdb",
    "postgresdb": {
      "database": "n8n",
      "host": "localhost",
      "port": "5432",
      "schema": "public",
      "user": "postgres",
      "password": "some-password"
    }
  }
}
```

And seeing this:

```
n.st

> n8n-monorepo@1.49.0 start /Users/ivov/Development/n8n
> run-script-os


> n8n-monorepo@1.49.0 start:default
> cd packages/cli/bin && ./n8n

User settings loaded from: /Users/ivov/.n8n/config
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
Loading config overwrites [ '/Users/ivov/.n8n/config-files/local-postgres.json' ]
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (BaseCommand)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(Use `node --trace-warnings ...` to show where the warning was created)
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (audit)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (execute)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (executeBatch)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (start)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (webhook)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (worker)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (db:revert)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (export:credentials)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (export:workflow)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (import:credentials)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (import:workflow)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (license:clear)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (license:info)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (list:workflow)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (ldap:reset)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (mfa:disable)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (update:workflow)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
(node:89674) Warning: Error
module: @oclif/core@4.0.7
task: findCommand (user-management:reset)
plugin: n8n
root: /Users/ivov/Development/n8n/packages/cli
message: configuration param 'database.type' not declared in the schema
configuration param 'database.postgresdb.database' not declared in the schema
configuration param 'database.postgresdb.host' not declared in the schema
configuration param 'database.postgresdb.port' not declared in the schema
configuration param 'database.postgresdb.schema' not declared in the schema
configuration param 'database.postgresdb.user' not declared in the schema
configuration param 'database.postgresdb.password' not declared in the schema
See more details with DEBUG=*
 ›   Error: command start not found
 ELIFECYCLE  Command failed with exit code 2.
```

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy committed:
fix existing N8N_CONFIG_FILES support

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

netroy commented:
The plan was to add support for `N8N_CONFIG_FILES` in the new config package in a new PR, but for now I've fixed the existing support. Created a new sub-issue ENG-132 to add support in the new package, were we can write tests more easily.

netroy committed:
"fix"

netroy committed:
Merge remote-tracking branch 'origin/master' into config-package

ivov commented on packages/@n8n/config/src/decorators.ts:
```diff
@@ -0,0 +1,79 @@
+		const classMetadata = globalMetadata.get(ConfigClass);
+		if (!classMetadata) {
+			// eslint-disable-next-line n8n-local-rules/no-plain-errors
+			throw new Error('Invalid config class: ' + ConfigClass.name);
```
Should we enable `n8n-local-rules/no-plain-errors` for this package? In this case it makes no difference as failing here breaks the build and is not reportable, but for future proofing any other errors we may need to throw in this package. Maybe the rule should be live at the top lint config file and be disabled on a case-by-case basis.

netroy replied:
I'd like to see if we can make the type-checker somehow catch these kind of errors, to prevent stuff like this even reaching runtime.

ivov commented on packages/cli/src/UserManagement/email/NodeMailer.ts:
```diff
@@ -1,87 +1,52 @@
-			ignoreTLS: !config.getEnv('userManagement.emails.smtp.startTLS'),
-		};
+	constructor(
+		globalConfig: GlobalConfig,
```
```suggestion
		readonly globalConfig: GlobalConfig,
```

netroy replied:
will update in the next PR

ivov commented on packages/cli/src/UserManagement/email/NodeMailer.ts:
```diff
@@ -1,87 +1,52 @@
+		globalConfig: GlobalConfig,
+		private readonly logger: Logger,
+	) {
+		const smtpConfig = globalConfig.userManagement.emails.smtp;
```
Can a consumer of `globalConfig` mutate it by accident? If so should we prevent this?

netroy replied:
right now, unfortunately yes. The plan is to switch to read-only proxies for the config eventually so that we can 
1. prevent writes
2. block any callers from outside our packages from reading any config

Created ENG-134 for that.

ivov commented on packages/cli/src/UserManagement/email/NodeMailer.ts:
Retested email sending and working.

ivov commented on packages/cli/src/UserManagement/email/UserManagementMailer.ts:
```diff
@@ -22,58 +21,37 @@ import { EventRelay } from '@/eventbus/event-relay.service';
 
 	constructor(
-		private readonly userRepository: UserRepository,
+		globalConfig: GlobalConfig,
```
```suggestion
		readonly globalConfig: GlobalConfig,
```

ivov commented on packages/cli/src/UserManagement/email/UserManagementMailer.ts:
```diff
@@ -22,58 +21,37 @@ import { EventRelay } from '@/eventbus/event-relay.service';
-			config.getEnv('userManagement.emails.mode') === 'smtp' &&
-			config.getEnv('userManagement.emails.smtp.host') !== '';
+		const emailsConfig = globalConfig.userManagement.emails;
+		this.isEmailSetUp = emailsConfig.mode === 'smtp' && emailsConfig.smtp.host !== '';
```
We could consider 'getters' for the new config.

netroy replied:
That is also a part of the plan, to have more utility methods/getters on the config classes.

ivov commented on packages/cli/src/databases/config.ts:
```diff
@@ -8,62 +8,64 @@ import type { PostgresConnectionOptions } from '@n8n/typeorm/driver/postgres/Pos
 	if (loggingOption) {
-		const optionsString = config.getEnv('database.logging.options').replace(/\s+/g, '');
-
+		const optionsString = loggingConfig.options.replace(/\s+/g, '');
```
Also in future we could have config doing its own validation.

netroy replied:
Yes. not only validation, but also transformation. so that JSON values could be auto-parsed, and we don't need to manually parse them, like we do for `credentials.overwrite.data`, and a few other properties.

ivov reviewed (approved):
🚀

janober commented:
Got released with [`n8n@1.50.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.50.0)
